array.h

```diff
@@ -1227,22 +1227,39 @@ class shape {
   }
 };
 
+namespace internal {
+
+template <size_t... DimIndices, class Shape, size_t... Extras>
+NDARRAY_HOST_DEVICE auto transpose_impl(const Shape& shape, index_sequence<Extras...>) {
+  return make_shape(
+      shape.template dim<DimIndices>()..., shape.template dim<sizeof...(DimIndices) + Extras>()...);
+}
+
+}  // namespace internal
+
 /** Create a new shape using a list of `DimIndices...` to use as the
  * dimensions of the shape. The new shape's i'th dimension will be the
  * j'th dimension of `shape` where j is the i'th value of `DimIndices...`.
+ * If i is not in `DimIndices...`, then i and j are equal.
  *
- * `transpose` requires `DimIndices...` to be a permutation, while
- * `reorder` accepts a list of indices that may be a subset of the
- * dimensions.
+ * `DimIndices...` to be a permutation of `[0, sizeof...(DimIndices))`,
  *
  * Examples:
  * - `transpose<2, 0, 1>(s) == make_shape(s.z(), s.y(), s.x())`
- * - `reorder<1, 2>(s) == make_shape(s.y(), s.z())` */
+ * - `transpose<1, 0>(s) == make_shape(s.y(), s.x(), s.z())` */
```

Comment:
Insert above: for a 3-dimensional Shape `s`:

```suggestion
 * - `transpose<1, 0>(s) == make_shape(s.y(), s.x(), s.z())` */
 For a 5-dimensional Shape `t`:
 * - `transpose<1, 0>(t) == make_shape(t.dim<1>(), t.dim<0>(), t.dim<2>(), t.dim<3>(), t.dim<4>());
 * - `transpose<2, 0>(t) == make_shape(t.dim<2>(), t.dim<0>(), t.dim<2>(), t.dim<3>(), t.dim<4>());  // this is probably not what you want
```

Reply:
Your second example is an error because it doesn't include 1, the DimIndices must be a permutation of [0, N) where N is the number of indices provided.

I think I'll make examples for 4D shapes, because I think the `.template dim` syntax is required, which is really verbose.